### ISSUE TITLE
Create a framework for Kitura template engines

### ISSUE THREAD

**Issue by vadimeisenbergibm** [IBM-Swift/Kitura#203]

**Comment by vadimeisenbergibm** [comment]
Added initial implementation in [Kitura Templating branch](https://github.com/IBM-Swift/Kitura/tree/Templating). See [Package.swift](https://github.com/IBM-Swift/Kitura/blob/Templating/Package.swift) and [main.swift](https://github.com/IBM-Swift/Kitura/blob/Templating/Sources/KituraSample/main.swift). Currently it works on OS X only.

See also [TemplateEngine protocol](https://github.com/vadimeisenbergibm/Kitura-TemplateEngine/blob/master/Sources/TemplateEngine.swift) and [its implementation for Mustache](https://github.com/vadimeisenbergibm/Kitura-MustacheTemplateEngine/tree/develop)

See package diagrams. The first diagram describes the current situation: 
![package_diagramvarianta](https://cloud.githubusercontent.com/assets/16784982/13432584/f611657c-dfd6-11e5-82d4-d00a0534a59b.PNG)
  The second diagram describes the situation in which Mustache package itself provides implementation of TemplateEngine protocol, so it can be used directly in Kitura applications. However, it adds a dependency from Mustache to Kitura-TemplateEngine package. 
![package_diagramvariantb](https://cloud.githubusercontent.com/assets/16784982/13432596/06c31c1c-dfd7-11e5-8af6-abd22589fe8c.PNG)

In both cases, any template engine can be integrated with Kitura by implementing TemplateEngine protocol, either as a separate package or inside the template engine package.

Any comments are welcome.

**Comment by rolivieri** [comment]
@vadimeisenbergibm I took a quick look at the diagram above and the protocol definition. I like that there is no direct dependency between Kitura-router and the actual template engine in both approaches. We leave the responsibility of creating the concrete instance of the template engine to the application. I think that’s great.

Now regarding which one of the two approaches, I am leaning towards approach # 1 since it does not require changes to the template engine library code. Instead, it is up to us (Kitura team) to provide the implementation for the protocol for the given template engine library. I think that’s the way to go; otherwise, we would need to ask each template engine owner to implement our protocol, which may not be feasible.

If I misunderstood anything, let me know!

**Comment by vadimeisenbergibm** [comment]
@rolivieri I am also in favor of approach # 1, which does not exclude the approach # 2 - any template engine owner can still add implementation of TemplateEngine protocol if they want.

**Comment by dfirsht** [comment]
I agree with Ricardo. We shouldn't be reliant on template engine owners implementing our protocol directly. So with approach 1, in order to add a new template engine, such as Stencil, we would create a Kitura-StencilTemplateEngine which would implement the TemplateEngine protocol with Stencil?

**Comment by vadimeisenbergibm** [comment]
@dfirsht Correct

**Comment by dfirsht** [comment]
Awesome, then I'm all in favor of approach 1.

**Comment by vadimeisenbergibm** [comment]
@rolivieri @dfirsht just to clarify approach 2 - we could submit a PR with our implementation to template engine owner. I am still in favor of approach 1.

**Comment by rfdickerson** [comment]
Yeah, I think it's great that template engine developers can choose either option. But if I were the developing Mustache or Stencil, I would of course, choose option 1. Since I would not want tight coupling on Kitura protocols. I think this adapter pattern sorts well to overlay the TemplateEngine protocol's implementation on Mustache.

**Comment by AnthonyMDev** [comment]
Is there a tutorial or guide on how to use template engines with Kitura out there yet? I haven't been able to find one. Thanks!

**Comment by vadimeisenbergibm** [comment]
Hi @AnthonyMDev. I am planning to write a tutorial by the end of May (on vacation now, so cannot do it sooner). The simplest way is to copy the sample from https://github.com/IBM-Swift/Kitura-Sample/tree/develop, **develop** branch. See https://github.com/IBM-Swift/Kitura-Sample/blob/develop/Sources/KituraSample/main.swift: lines 24, 149-176.

**Comment by AnthonyMDev** [comment]
@vadimeisenbergibm Thanks! I looked at those a bit and that makes sense. As soon as we get an ORM, this framework is going to be amazing. 👍 I'm considering looking into building one myself, but I wasn't sure if that's in the works yet from you guys?
